File: backend/app/models/task.py
from datetime import date, datetime
from enum import Enum

from sqlalchemy import Boolean, Date, DateTime, Float, ForeignKey, Integer, JSON, String, UniqueConstraint, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class TaskStatus(str, Enum):
    TODO = "TODO"
    IN_PROGRESS = "IN_PROGRESS"
    IN_REVIEW = "IN_REVIEW"
    REJECTED = "REJECTED"
    DONE = "DONE"
    CLOSED = "CLOSED"
    DELETED = "DELETED"


class TaskType(str, Enum):
    GENERAL = "GENERAL"
    DOCUMENT = "DOCUMENT"
    CODE = "CODE"


class TaskPriority(str, Enum):
    URGENT = "URGENT"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class TaskParticipantRole(str, Enum):
    PARTICIPANT = "PARTICIPANT"


class Task(Base):
    __tablename__ = "tasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    project_id: Mapped[int] = mapped_column(ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True)
    column_id: Mapped[int] = mapped_column(ForeignKey("board_columns.id"), nullable=False, index=True)
    owner_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(160), nullable=False)
    description: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    task_type: Mapped[str] = mapped_column(String(32), nullable=False, default=TaskType.GENERAL.value)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default=TaskStatus.TODO.value, index=True)
    priority: Mapped[str] = mapped_column(String(32), nullable=False, default=TaskPriority.MEDIUM.value)
    due_date: Mapped[date | None] = mapped_column(Date, nullable=True, index=True)
    labels: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    is_blocked: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, index=True)
    current_blocker_summary: Mapped[str | None] = mapped_column(String(500), nullable=True)
    acceptance_summary: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    participants: Mapped[list["TaskParticipant"]] = relationship(
        back_populates="task",
        cascade="all, delete-orphan",
        order_by="TaskParticipant.created_at",
        passive_deletes=True,
    )
    subtasks: Mapped[list["Subtask"]] = relationship(
        back_populates="task",
        cascade="all, delete-orphan",
        order_by="Subtask.position",
        passive_deletes=True,
    )
    dependencies: Mapped[list["TaskDependency"]] = relationship(
        foreign_keys="TaskDependency.task_id",
        back_populates="task",
        cascade="all, delete-orphan",
        passive_deletes=True,
    )
    work_logs: Mapped[list["WorkLog"]] = relationship(
        back_populates="task",
        cascade="all, delete-orphan",
        order_by="WorkLog.work_date.desc(), WorkLog.created_at.desc()",
        passive_deletes=True,
    )


class TaskParticipant(Base):
    __tablename__ = "task_participants"
    __table_args__ = (UniqueConstraint("task_id", "user_id", name="uq_task_participants_task_user"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    role: Mapped[str] = mapped_column(String(32), nullable=False, default=TaskParticipantRole.PARTICIPANT.value)
    removed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    task: Mapped[Task] = relationship(back_populates="participants")


class Subtask(Base):
    __tablename__ = "subtasks"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True)
    title: Mapped[str] = mapped_column(String(160), nullable=False)
    is_completed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    completed_by_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True, index=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    position: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    task: Mapped[Task] = relationship(back_populates="subtasks")


class TaskDependency(Base):
    __tablename__ = "task_dependencies"
    __table_args__ = (UniqueConstraint("task_id", "depends_on_task_id", name="uq_task_dependencies_pair"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True)
    depends_on_task_id: Mapped[int] = mapped_column(
        ForeignKey("tasks.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    task: Mapped[Task] = relationship(foreign_keys=[task_id], back_populates="dependencies")


class WorkLog(Base):
    __tablename__ = "work_logs"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    task_id: Mapped[int] = mapped_column(ForeignKey("tasks.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    work_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    hours: Mapped[float] = mapped_column(Float, nullable=False)
    content: Mapped[str] = mapped_column(String(2000), nullable=False)
    work_type: Mapped[str] = mapped_column(String(50), nullable=False, default="GENERAL")
    is_blocked: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False, index=True)
    blocked_reason: Mapped[str | None] = mapped_column(String(500), nullable=True)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    resolved_by_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True, index=True)
    resolution_note: Mapped[str | None] = mapped_column(String(500), nullable=True)
    commit_hash: Mapped[str | None] = mapped_column(String(120), nullable=True)
    branch_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    repository_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    git_synced: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True, index=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    task: Mapped[Task] = relationship(back_populates="work_logs")

File: frontend/src/types/task.ts
import type { UserPublic } from "./auth";
import type { BoardColumn } from "./project";

export type TaskStatus = "TODO" | "IN_PROGRESS" | "IN_REVIEW" | "REJECTED" | "DONE" | "CLOSED" | "DELETED";
export type TaskType = "GENERAL" | "DOCUMENT" | "CODE";
export type TaskPriority = "URGENT" | "HIGH" | "MEDIUM" | "LOW";

export interface TaskParticipant {
  id: number;
  task_id: number;
  user_id: number;
  role: "PARTICIPANT";
  removed_at: string | null;
  created_at: string;
  user?: UserPublic | null;
}

export interface Subtask {
  id: number;
  task_id: number;
  title: string;
  is_completed: boolean;
  completed_by_id: number | null;
  completed_at: string | null;
  position: number;
  created_at: string;
  updated_at: string;
}

export interface TaskDependency {
  id: number;
  task_id: number;
  depends_on_task_id: number;
  created_at: string;
}

export interface WorkLog {
  id: number;
  task_id: number;
  user_id: number;
  work_date: string;
  hours: number;
  content: string;
  work_type: string;
  is_blocked: boolean;
  blocked_reason: string | null;
  resolved_at: string | null;
  resolved_by_id: number | null;
  resolution_note: string | null;
  commit_hash: string | null;
  branch_name: string | null;
  repository_url: string | null;
  git_synced: boolean;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskBlockerSummary {
  is_blocked: boolean;
  current_blocker_summary: string | null;
  unresolved_count: number;
}

export interface TaskBase {
  id: number;
  project_id: number;
  column_id: number;
  owner_id: number;
  title: string;
  description: string | null;
  task_type: TaskType;
  status: TaskStatus;
  priority: TaskPriority;
  due_date: string | null;
  labels: string[];
  progress: number;
  is_blocked: boolean;
  current_blocker_summary: string | null;
  acceptance_summary: string | null;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskBoardCard extends TaskBase {
  owner?: UserPublic | null;
  participants: TaskParticipant[];
  subtask_total: number;
  subtask_completed: number;
  latest_work_log_at: string | null;
  blocker_summary?: TaskBlockerSummary | null;
}

export interface TaskDetail extends TaskBase {
  owner?: UserPublic | null;
  column?: BoardColumn | null;
  participants: TaskParticipant[];
  subtasks: Subtask[];
  dependencies: TaskDependency[];
  work_logs: WorkLog[];
  blocker_summary?: TaskBlockerSummary | null;
}

export interface TaskCreatePayload {
  title: string;
  description?: string | null;
  task_type?: TaskType;
  priority?: TaskPriority;
  due_date?: string | null;
  owner_id: number;
  participant_ids?: number[];
  column_id?: number | null;
  status?: TaskStatus;
  labels?: string[];
}

export interface TaskUpdatePayload {
  title?: string;
  description?: string | null;
  task_type?: TaskType;
  priority?: TaskPriority;
  due_date?: string | null;
  owner_id?: number;
  column_id?: number | null;
  labels?: string[];
}

export interface WorkLogCreatePayload {
  work_date: string;
  hours: number;
  content: string;
  work_type?: string;
  is_blocked?: boolean;
  blocked_reason?: string | null;
  commit_hash?: string | null;
  branch_name?: string | null;
  repository_url?: string | null;
}

export interface WorkLogUpdatePayload extends Partial<WorkLogCreatePayload> {}

export interface BlockerResolvePayload {
  resolution_note: string;
}
